x-pack/plugin/rest-test/src/main/java/org/elasticsearch/xpack/resttest/resthandler/ExampleCatAction.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.xpack.resttest.resthandler;
+
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.search.SearchScrollRequest;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.Table;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.cat.AbstractCatAction;
+import org.elasticsearch.rest.action.cat.RestTable;
+import org.elasticsearch.search.SearchHit;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+
+/**
+ * Example of adding a cat action with a plugin.
+ */
+
+// TODO: how to extract this plugin into a separate repository
+// and how to integrate the separate plugin repository together
+// with the elasticsearch repository.
+public class ExampleCatAction extends AbstractCatAction {
+
+    ExampleCatAction(final RestController controller) {
+        controller.registerHandler(GET, "/_cat/example", this);
+        controller.registerHandler(POST, "/_cat/example", this);
+    }
+
+    @Override
+    public String getName() {
+        return "rest_handler_cat_example";
+    }
+
+    @Override
+    protected RestChannelConsumer doCatRequest(final RestRequest request, final NodeClient client) {
+        Iterator<IndexMetaData> indices = client.admin().cluster().prepareState().get().getState().getMetaData().getIndices().valuesIt();
```

Comment:
This `prepareState().get()` isn't allowed on the main Rest handler thread. It would cause the Rest server to block, you need to do this with a listener callback.

Something like:
```suggestion
    client.admin().cluster().prepareState().execute(new RestResponseListener<ClusterStateResponse>(channel) {
        public RestResponse buildResponse(ClusterStateResponse response) throws Exception {
            Table table = buildTable(response.getState().getMetaData().getIndices());
             return RestTable.buildResponse(table, channel);
        }
     });
```
And then break all the table building code into a new `buildTable(ImmutableOpenMap<String, IndexMetaData> indices)` method.